docs/adr/0009_controller_v2_lib_migration.md

```diff
@@ -0,0 +1,406 @@
+
+To support the migration strategy, the first phase will require a careful rearchitecture of the critical controllers.  
+In the current controller architecture, the fetching and resolution of component versions is performed directly within reconciliation.
+
+This is problematic for two reasons:
+
+1. Resolution is tightly coupled to the reconciliation loop, which reduces throughput and introduces blocking.
+2. The same resolution logic cannot be easily reused across multiple controllers, resulting in duplicated effort and inconsistent behavior.
+
+### Proposed Solution: Component Version Resolution Service
+
+We propose introducing a dedicated **Component Version Resolution Service** as the substrate for all controllers.  
+This service encapsulates resolution logic behind a consistent interface and provides caching, deduplication, and concurrency control.
+
+#### Responsibilities
+
+* **On-demand resolution** of component versions (keyed by repository, component name, and version).
+* **Result caching** with a bounded LRU policy to avoid repeated expensive lookups.
+* **Deduplication** of concurrent requests for the same key using singleflight semantics.
+* **Concurrency limiting** with a worker pool to bound resource usage when resolving large volumes of components.
+* **Reusable API** that can be called from multiple controllers without embedding resolution logic.
+
+#### Architecture of Core Component Version Resolution Logic
+
+```mermaid
+flowchart TD
+    subgraph Controllers
+        C[ComponentVersion Controller]
+        Rc[Resource Controller]
+    end
+
+    subgraph Service[Resolution Service]
+        LRU[Cache of Resolved Versions k8s.io/utils/lru]
+
+        Q[Add Lookup Request]
+        subgraph Queue[Version Lookup Queue]
+            SF[golang.org/x/sync/singleflight]
+            SEM[bounded channel]
+        end
+
+        subgraph POOL[Version Lookup Worker Pool]
+            WORKER1[Worker 1]
+            WORKER2[Worker 2]
+            WORKER3[Worker 3]
+        end
+    end
+
+    subgraph Library[New OCM Reference Library]
+        subgraph Repository[bindings/go/repository]
+            RESOLVER[Resolver]
+        end
+    end
+
+%% Controller lookups
+    C -. ask for version .-> LRU
+    Rc -. ask for version .-> LRU
+
+%% Cache miss triggers async path
+    LRU -. miss .-> Q
+    Q -. deduplicate in flight requests .-> SF
+    SF -. limit concurrent lookups .-> SEM
+    SEM --> WORKER1
+    SEM --> WORKER2
+    SEM --> WORKER3
+
+    WORKER1 <-. Resolve and Verify .-> RESOLVER
+    WORKER2 <-. Resolve and Verify .-> RESOLVER
+    WORKER3 <-. Resolve and Verify .-> RESOLVER
+
+    WORKER3 -. resolved version .-> LRU
+```
+
+Explanation of flowchart components:
+
+- Controllers always hit the LRU cache first.
+- On a cache miss, the request is added to the lookup queue, and status is set to "in progress"
+- singleflight deduplicates requests for the same version to ensure only one actual backend fetch occurs.
```

Comment:
I assume requests from the SAME object, right? Because there could be requests for the same thing but from a different source.